Mathlib/NumberTheory/ModularForms/Bounds.lean

```diff
@@ -0,0 +1,331 @@
+
+open Complex ModularFormClass
+
+/-- General result on bounding q-expansion coefficients using a bound on the norm of the function.
+This will get used twice over, once for cusp forms (with `e = k / 2`) and once for modular forms
+(with `e = k`). -/
+lemma qExpansion_coeff_isBigO_of_norm_isBigO {k : ℤ} {Γ : Subgroup (GL (Fin 2) ℝ)}
+    [Γ.IsArithmetic] {F : Type*} [FunLike F ℍ ℂ] [ModularFormClass F Γ k] (f : F) (e : ℝ)
+    (hF : IsBigO (comap UpperHalfPlane.im (𝓝 0)) f (fun τ ↦ τ.im ^ (-e))) :
+    (fun n ↦ (qExpansion Γ.strictWidthInfty f).coeff n) =O[atTop] fun n ↦ (n : ℝ) ^ e := by
+  let h := Γ.strictWidthInfty
+  have hh : 0 < h := Γ.strictWidthInfty_pos_iff.mpr Fact.out
+  haveI : NeZero h := ⟨hh.ne'⟩
+  have hΓ : h ∈ Γ.strictPeriods := Γ.strictWidthInfty_mem_strictPeriods
+  obtain ⟨C, Cpos, hC⟩ := hF.exists_pos
+  rw [isBigO_iff]
+  rw [IsBigOWith, eventually_comap] at hC
+  use (1 / Real.exp (-2 * Real.pi / ↑h)) * C
+  have := Tendsto.eventually (tendsto_inv_atTop_zero.comp tendsto_natCast_atTop_atTop) hC
+  filter_upwards [eventually_gt_atTop 0, this] with n hn hn'
+  rw [qExpansion_coeff_eq_intervalIntegral (t := 1 / n) f hh hΓ _ (by positivity),
+    ← intervalIntegral.integral_const_mul]
+  simp only [ofReal_div, ofReal_one, ofReal_natCast]
+  refine intervalIntegral.norm_integral_le_integral_norm (by positivity) |>.trans ?_
+  let F (x : ℝ) : ℝ := ‖1 / ↑h * (1 / 𝕢 h ((x : ℂ) + 1 / n * I) ^ n
+      * f ⟨(x : ℂ) + 1 / n * Complex.I, by simp [hn]⟩)‖
+  have (x : ℝ) : F x ≤ 1 / h * ((1 / Real.exp (-2 * Real.pi / ↑h))) * (C * n ^ e) := by
+    simp only [F, norm_mul, norm_mul, norm_div, norm_real, norm_one, norm_div, norm_one, norm_pow,
+      mul_assoc, Real.norm_of_nonneg hh.le]
+    refine mul_le_mul_of_nonneg_left (mul_le_mul ?_ ?_ ?_ ?_) ?_
+    · rw [Function.Periodic.norm_qParam, add_im, ofReal_im, zero_add, mul_I_im, ← ofReal_one,
+        ← ofReal_natCast, ← ofReal_div, ofReal_re, mul_one_div, div_right_comm, ← Real.exp_nat_mul,
+        mul_div_cancel₀ _ (mod_cast hn.ne')]
+    · refine (hn' _ ?_).trans (le_of_eq ?_) <;>
+        simp_rw [← UpperHalfPlane.coe_im, UpperHalfPlane.coe_mk_subtype, add_im, ofReal_im,
+          zero_add, mul_I_im, ← ofReal_one, ← ofReal_natCast, ← ofReal_div, ofReal_re]
+      · simp
+      · rw [one_div, Real.rpow_neg_eq_inv_rpow, inv_inv, Real.norm_of_nonneg (by positivity)]
+    all_goals positivity
+  refine (intervalIntegral.integral_mono (by positivity) ?_ ?_ this).trans (le_of_eq ?_)
+  · refine continuous_const.mul (.mul ?_ ?_) |>.norm |>.intervalIntegrable _ _
+    · simp_rw [Function.Periodic.qParam, ← Complex.exp_nat_mul, one_div, ← Complex.exp_neg]
+      fun_prop
+    · exact (continuous f).comp (by fun_prop)
+  · exact continuous_const.intervalIntegrable ..
+  · rw [intervalIntegral.integral_const, sub_zero, smul_eq_mul]
+    simp [← mul_assoc, mul_inv_cancel₀ (NeZero.ne (h : ℝ)),
+      Real.norm_of_nonneg (show 0 ≤ (n : ℝ) ^ e by positivity)]
```

Comment:
```suggestion
    simp [field, Real.norm_of_nonneg (show 0 ≤ (n : ℝ) ^ e by positivity)]
```
More simplifications may be possible in this proof, but I'm getting too tired...